<a href="https://colab.research.google.com/github/byui-cse/cse380-notebooks/blob/master/10_3_About_Patterns_and_Probabilities.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# About Patterns and Probabilities
## Class Directed Learning
### Due: Tuesday, 9 March 2021, 11:59 pm

## TODO Explore and Wonder

Regarding spanning trees of ladder graphs:

What is the probability ($\lim_{n \rightarrow \infty} P(n)$) that a randomly-selected spanning tree of an $n$-rung ladder graph contains the bottom rung?

| n | P(n) |
|---|------|
| 1 |      |
| 2 |      |
| 3 |      |
| 4 |      |
| 5 |      |


### Recurrence Relations

In [6]:
#https://docs.python.org/3/library/functools.html
from functools import cache


def get_num_edges(n):
    return n + 2 * (n - 1)

# Num of with BR given n
@cache
def f(n):
    if(n < 2):
        return 1
    return 4 * f(n-1) - f(n-2)

@cache
def g(n):
    if(n < 2):
        return n
    return 4 * g(n-1) - g(n-2)

In [7]:
g_nums_sum = sum([*map(g, range(1,100))])
f_nums_sum = sum([*map(f, range(1,100))])

# bottom rungs / spanning trees
f_nums_sum/ g_nums_sum

0.7320508075688773

About 73% 

Let $f(n) =$ NSTIBR$(n)$:

$f(n) = 4f(n-1) - f(n-2)$ for $n > 1$;

$f(0) = 1$,

$f(1) = 1$.

Let $g(n) =$ NST$(n)$:

$g(n) = 4g(n-1) - g(n-2)$ for $n > 1$;

$g(0) = 0$,

$g(1) = 1$.

#### TODO Check Recurrences

In [8]:
from functools import cache

@cache
def f(n):
    if(n < 2):
        return 1
    return 4 * f(n-1) - f(n-2)

for x in range(40):
    print(f(x))

1
1
3
11
41
153
571
2131
7953
29681
110771
413403
1542841
5757961
21489003
80198051
299303201
1117014753
4168755811
15558008491
58063278153
216695104121
808717138331
3018173449203
11263976658481
42037733184721
156886956080403
585510091136891
2185153408467161
8155103542731753
30435260762459851
113585939507107651
423908497265970753
1582048049556775361
5904283700961130691
22035086754287747403
82236063316189858921
306909166510471688281
1145400602725696894203
4274693244392315888531


In [9]:
from functools import cache

@cache
def g(n):
    if(n < 2):
        return n
    return 4 * g(n-1) - g(n-2)

for x in range(6):
    print(g(x))

0
1
4
15
56
209


Check the calculations in this table (maybe write recursive functions) to verify they are accurate.

Do they match what you found yesterday in your DPC?

| n | f(n) | f(n-1) | 4f(n-1)  | f(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    1 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      1 |        4 |      N/A |  N/A |
| 2 |    3 |      1 |        4 |        1 |    3 |
| 3 |   11 |      3 |       12 |        1 |   11 |
| 4 |   41 |     11 |       44 |        3 |   41 |
| 5 |  153 |     41 |      164 |       11 |  153 |

| n | g(n) | g(n-1) | 4g(n-1)  | g(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    0 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      0 |        0 |      N/A |  N/A |
| 2 |    4 |      1 |        4 |        0 |    4 |
| 3 |   15 |      4 |       16 |        1 |   15 |
| 4 |   56 |     15 |       60 |        4 |   56 |
| 5 |  209 |     56 |      224 |       15 |  209 |

### TODO Find closed-form formulas

Can you find closed-form formulas for $f(n)$ and $g(n)$?

A closed-form formula expressing these functions in terms of operations on $n$, **without** referring to previous calculated values of the functions.

#### Hint:

http://www.ist.tugraz.at/aichholzer/teaching/eca/spanning_trees_in_ladders.pdf

#### Possibly Illuminating Calculations

Remember continued fractions?

In [10]:
from fractions import Fraction as frac

def contfrac2frac(seq):
    """Convert the simple continued fraction in `seq`
       into a fraction with numerator num and denominator den.
    """
    num, den = 1, 0
    for u in reversed(seq):
        num, den = den + num * u, num
    return frac(num, den)

def frac2contfrac(f):
    """Build the simple continued fraction expansion of fraction f.
    """
    seq = []
    frac2contfrac_rec(f, seq)
    return seq

def frac2contfrac_rec(f, seq):
    n = f.numerator
    d = f.denominator
    if d != 0:
        seq.append(n // d)
        if n % d != 0:
            frac2contfrac_rec(frac(d, n % d), seq)

def eval_frac(f):
    """Evaluate the fraction f as a float.
    """
    return f.numerator / f.denominator

In [11]:
from math import sqrt

value = sqrt(3) - 1
value_as_cf = frac2contfrac(frac.from_float(value))[:21]
cf_to_value = contfrac2frac(value_as_cf)
print(value, value_as_cf, cf_to_value)
print(eval_frac(cf_to_value))

0.7320508075688772 [0, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2] 302632/413403
0.732050807565499


In [12]:
from pprint import pprint

pprint(list(map(lambda n: contfrac2frac(value_as_cf[:n]), range(20, 5, -1))))

[Fraction(110771, 151316),
 Fraction(81090, 110771),
 Fraction(29681, 40545),
 Fraction(21728, 29681),
 Fraction(7953, 10864),
 Fraction(5822, 7953),
 Fraction(2131, 2911),
 Fraction(1560, 2131),
 Fraction(571, 780),
 Fraction(418, 571),
 Fraction(153, 209),
 Fraction(112, 153),
 Fraction(41, 56),
 Fraction(30, 41),
 Fraction(11, 15)]
